[pki] Add possibility of renewing certificates #83

NoumbissiValere opened the pull request:
Added a new function in the Ca model which
renews the certificate, private key and
validity end date by calling the generate
function and updating the validity end date.
Also added actions in CA and cert admins to
make use of this newly added function.
Also added a confirmation page to the actions
which Fixes #85

Fixes #83

NoumbissiValere commented:
@nemesisdesign Please can you review this and give me feedback?
Thanks

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -119,6 +120,16 @@ def crl_view(self, request, pk):
+        for ca in queryset:
+            ca.renew_ca()
+            renewed_rows += 1
+        message = _('{0} CAs have been successfully renewed'.format(renewed_rows))
```
please use pluralization to display the correct message depending on the amount of objects on which the operation was performed, see here: https://docs.djangoproject.com/en/3.0/topics/i18n/translation/#pluralization

Follow the exact style used in that example.

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -175,6 +186,16 @@ def revoke_action(self, request, queryset):
+        for cert in queryset:
+            cert.ca.renew_ca()
+            renewed_rows += 1
+        message = _('{0} Certificates have been successfully renewed'.format(renewed_rows))
```
please use pluralization to display the correct message depending on the amount of objects on which the operation was performed, see here: https://docs.djangoproject.com/en/3.0/topics/i18n/translation/#pluralization

Follow the exact style used in that example.

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -175,6 +186,16 @@ def revoke_action(self, request, queryset):
+    def renew_cert(self, request, queryset):
+        renewed_rows = 0
+        for cert in queryset:
+            cert.ca.renew_ca()
```
why are you renewing the CA? In this case we have to renew only the certificate.

NoumbissiValere commented:
Hello @nemesisdesign I have updated this PR but `travis ci` has not gone through it, don't know why. Please can you review it and give me feedback. if there are errors, will update the PR ASAP. 
Thanks

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -429,6 +429,23 @@ def _add_extensions(self, cert):
+            self.ca.validity_end = default_ca_validity_end()
+            self.ca.serial_number = uuid.uuid4().int
+            self.ca.save()
+            self.validity_end = default_cert_validity_end()
```
wait a minute. If the model has the `ca` attribute it means we're dealing with a certificate.

So if the certificate is being renewed, the renewal of the CA is triggered.
If my understanding of this code is correct, the implementation is wrong.

When renewing a certificate, only the single certificate should be renewed.
When renewing a CA, all the certificates related to it should be renewed.

I would keep this private method, but make sure it only acts on `self`. See below for futher suggestions.

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -449,6 +466,13 @@ def get_revoked_certs(self):
+        Renew the certificate, private key and
+        validity end date of a CA
+        """
+        self._renew()
```
I would rename this method to just `renew`, after renewing the CA (calling `super()`), I would loop over the related certificates and call renew on each of them.

Since the second operation could take potentially very long, I would implement it in a separate method, so that in OpenWISP we can then overwrite the method and move it to a celery task if needed.

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -429,6 +429,23 @@ def _add_extensions(self, cert):
             ])
         return cert
 
+    def _renew(self):
```
I'd call this just `renew`

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -499,3 +523,10 @@ def revoke(self):
         self.revoked_at = now
         self.save()
+
+    def renew_cert(self):
```
I'd remove this since it's defined on the base model

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -429,6 +429,23 @@ def _add_extensions(self, cert):
+        else:
+            self.validity_end = default_ca_validity_end()
+        self._generate()
+        self.serial_number = uuid.uuid4().int
```
the instructions to generate a new serial number is already present in the code and is being repeated here, hence it's better to move this to a private method (eg: `_generate_serial_number()`) to ensure future maintainers won't accidentally forget to upgrade part of the code if needed

nemesisdesign commented on django_x509/tests/test_ca.py:
```diff
@@ -622,3 +622,15 @@ def test_datetime_to_string(self):
                          utc_datetime.strftime(utc_time))
+
+    def test_renew_certificate_key(self):
+        ca = self._create_ca()
```
please generate 2 certs related to the CA and ensure that when the renew is triggered, also the related certs are renewed

nemesisdesign commented on django_x509/tests/test_ca.py:
```diff
@@ -622,3 +622,15 @@ def test_datetime_to_string(self):
         self.assertEqual(datetime_to_string(utc_datetime),
                          utc_datetime.strftime(utc_time))
+
+    def test_renew_certificate_key(self):
```
test method name can be just `test_renew`, keep it simple

nemesisdesign commented on django_x509/tests/test_cert.py:
```diff
@@ -488,3 +488,23 @@ def test_generate_ca_with_passphrase(self):
         ca.save()
         self.assertIsInstance(ca.pkey, crypto.PKey)
+
+    def test_renew_certificate_and_key(self):
```
test method name can be just `test_renew`, keep it simple

nemesisdesign commented on django_x509/tests/test_cert.py:
```diff
@@ -488,3 +488,23 @@ def test_generate_ca_with_passphrase(self):
+        cert = self._create_cert()
+        old_cert = cert.certificate
+        old_key = cert.private_key
+        old_validity_end = cert.validity_end
```
ca should not be changed.

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -119,6 +120,21 @@ def crl_view(self, request, pk):
+            ca.renew_ca()
+            renewed_rows += 1
+        message = ngettext(
+            '%(renewed_rows)d CA has been successfully renewed',
```
I think we could make more explicit the fact that also related certificates have been renewed, eg:

`CA and its related certificates have been successfully renewed`, maybe in this form we don't need to pluralize.

I also thought it would be good to have a confirmation step, but that can be done afterwards, so I created #85

NoumbissiValere commented:
Thank you for taking out time to review and give me feedback. I think I now see what you have been seeing. I will update the PR ASAP.  And also add the confirmation page.

NoumbissiValere commented:
Good day @nemesisdesign I have updated this PR and also added the confirmation page to handle #85 . while testing the work on the UI with 3 CAs having 3 related certificates, i noticed the admin UI was a little slow and when i think of circumstances where an admin can have 10+ relations, it will really be slow so i was thinking of adding the celery task too to make renewing of related certificates to happen at the background. 
Please review the work and tell me if i should add the celery task now or leave it for later.

Thanks

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -175,6 +207,27 @@ def revoke_action(self, request, queryset):
+        count = len(queryset)
+        if request.POST.get('post'):
+            self.model.renew_certs(queryset)
+            renewed_rows = count
```
```python
renewed_rows = 0
for cert in queryset:
    cert.renew()
    renewed_rows += 1
```

Keep consistency with what you did in the cert action.

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -175,6 +207,27 @@ def revoke_action(self, request, queryset):
     revoke_action.short_description = _('Revoke selected certificates')
 
+    def renew_cert(self, request, queryset):
+        count = len(queryset)
```
remove this line

NoumbissiValere commented on requirements.txt:
```diff
@@ -3,3 +3,4 @@ django-model-utils>=4.0
 jsonfield>=3.1.0,<4.0.0
 cryptography>=2.4.0,<2.9.0
 pyopenssl>=17.5.0,<20.0.0
+openwisp_utils>=0.4.0,<0.5.0
```
I added these because the UI of `django_x509` was not like the other `openwisp module's`. It was still having plain django's UI. I wanted to look like the other modules. But since it was not requested, i will remove it

NoumbissiValere commented:
@nemesisdesign I have updated the PR please can you review it and give me feedback?
Thanks

nemesisdesign commented on django_x509/base/admin.py:
```diff
@@ -73,6 +74,35 @@ def get_fields(self, request, obj=None):
+        context = dict()
+        if ca_count:
+            message = _('Are you sure you want to renew the selected CAs and their related certificates? '
+                        'All the CAs listed below with related certificates will be renewed.')
```
there's no reason to put this here, can you put it back in the template and use `blocktranslate` if needed please?
use <br> to separate the two lines.

nemesisdesign commented on django_x509/base/models.py:
```diff
@@ -429,6 +429,18 @@ def _add_extensions(self, cert):
+        if hasattr(self, 'ca'):
+            self.validity_end = default_cert_validity_end()
+        else:
+            self.validity_end = default_ca_validity_end()
```
have you tried with: 

```
self.validity_end = self.__class__().validity_end
```

It would eliminate the need of the if.

nemesisdesign commented on django_x509/templates/admin/django_x509/renew_confirmation.html:
```diff
@@ -0,0 +1,55 @@
+            <input type="hidden" name="post" value="yes">
+            <input type="submit" value="Yes I'm sure">
+            <input type="hidden" name="action" value="{{ action }}">
+            <a href="{% url 'admin:'|add:cancel_url %}" class="button cancel-link"> No, take me back </a>
```
There's an extra space here before ` No,`

NoumbissiValere commented:
@nemesisdesign I have updated this PR as required
